cli: add --http-addr flag

glycerine opened the pull request:
For security, the --http-addr flags makes the Admin UI
bind an IP address that can be private or distinct from
the cluster (such as 127.0.0.1).

Fixes #7474.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7475)

<!-- Reviewable:end -->

glycerine committed:
add --http-addr flag

For security, the --http-addr flags makes the Admin UI
bind an IP address that can be private or distinct from
the cluster. If --host is given, we default to that
address if --http-addr is unspecified.

Fixes #7474.

glycerine commented:
From the CI test failures, it looks like urls such as https://127.0.0.1:32821/_status/gossip/local  also got shifted to port 8080. I'm trying to figure this out.

knz commented:
@tamird @mberhault what's the best place to add a test for this flag in our acceptance tests?

mberhault commented on cli/cliflags/names.go:
```diff
@@ -33,6 +33,7 @@ const (
 	PasswordName          = "password"
 	PortName              = "port"
 	HTTPPortName          = "http-port"
+	HTTPIPAddrName        = "http-addr"
```
small nit: use `HTTPAddrName`, it doesn't have to be an IP address, dns names work fine.

mberhault replied:
Actually, a bigger nit: let's stick to the same naming scheme as `{host, port}`, where we now have `{http-host, http-port}`.

mberhault commented on cli/flags.go:
```diff
@@ -92,6 +92,9 @@ Database server port to connect to.`),
 Database server port to connect to for HTTP requests.`),
 
+	forClient(cliflags.HTTPIPAddrName): wrapText(`
+Database server IP address to connect to for HTTP requests.`),
```
Ditto for the help message: `Database server port to connect to for HTTP requests.`

tamird commented:
Please hold off on merging this - it's going to conflict with multiple addresses / external address work, and the semantics aren't completely clear (i.e. we may want to move to specifying full addresses rather than addresses and ports).

mberhault commented:
A test would be nice, and the acceptance tests are probably the best place since we would ideally like some networking isolation.

@tamird: thoughts on the tests? Also how this may fit in with future post modifications?

glycerine commented:
Marc - I agree a test or two would be nice. I had the worst time trying to get the acceptance suite to run locally. I filed https://github.com/cockroachdb/cockroach/issues/7477 about that.

Tamir - okay, let me know when I might be of assistance.

glycerine commented:
@tamird Have the semantics been decided? I'd like to move forward on this.

Working at adding a test and fixing the nits above...

glycerine commented:
I've added a surgically precise test unit test.  I think an acceptance test is beyond the scope of what I, with my limited resources, can attempt at this point.

Also, I've addressed the nits.

@tamird, @mberhault, @knz, would you be okay with proceeding? It's a very small change, and it fixes a very big security hole. This would unblock me from serious further testing.

If there's a big change later towards combined IPs and ports, this really doesn't add much more to change.

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7475#-:-KMEVerb73KwQJtuemaa:1417088721)*
<!-- Sent from Reviewable.io -->

knz commented:
@glycerine I'll take that one. The fact we want to do things perhaps in a different way in the future does not prevent us from accepting your patch now, for use in the interim.

glycerine commented:
Thanks @knz.

glycerine commented:
@jseldess Now that https://github.com/cockroachdb/docs/pull/443 merged, I think the docs-todo label can  be removed.

jseldess commented:
Thanks, @glycerine. Just tweaking a few minor things before publishing (probably early tomorrow).
